presto-tests/src/main/java/com/facebook/presto/tests/statistics/MetricComparator.java

```diff
@@ -13,88 +13,108 @@
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.joining;
 
 public final class MetricComparator
 {
     private MetricComparator() {}
 
-    public static List<MetricComparison> createMetricComparisons(Plan queryPlan, Map<PlanNodeId, PlanNodeStatsEstimate> estimates, StageInfo outputStageInfo)
+    public static List<MetricComparison> getMetricComparisons(String query, DistributedQueryRunner runner, List<Metric> metrics)
     {
-        return Stream.of(Metric.values()).flatMap(metric -> {
-            Map<PlanNodeId, PlanNodeStatsEstimate> actuals = extractActualStats(outputStageInfo);
-            return estimates.entrySet().stream().map(entry -> {
-                // todo refactor to stay in PlanNodeId domain ????
-                PlanNode node = planNodeForId(queryPlan, entry.getKey());
-                PlanNodeStatsEstimate estimate = entry.getValue();
-                Optional<PlanNodeStatsEstimate> execution = Optional.ofNullable(actuals.get(node.getId()));
-                return createMetricComparison(metric, node, estimate, execution);
-            });
-        }).collect(Collectors.toList());
-    }
+        List<OptionalDouble> estimatedValues = getEstimatedValues(metrics, query, runner);
+        List<OptionalDouble> actualValues = getActualValues(metrics, query, runner);
 
-    private static PlanNode planNodeForId(Plan queryPlan, PlanNodeId id)
-    {
-        return searchFrom(queryPlan.getRoot())
-                .where(node -> node.getId().equals(id))
-                .findOnlyElement();
+        ImmutableList.Builder<MetricComparison> metricComparisons = ImmutableList.builder();
+        for (int i = 0; i < metrics.size(); ++i) {
+            //noinspection unchecked
+            metricComparisons.add(new MetricComparison(
+                    metrics.get(i),
+                    estimatedValues.get(i),
+                    actualValues.get(i)));
+        }
+        return metricComparisons.build();
     }
 
-    private static Map<PlanNodeId, PlanNodeStatsEstimate> extractActualStats(StageInfo outputStageInfo)
+    private static List<OptionalDouble> getEstimatedValues(List<Metric> metrics, String query, DistributedQueryRunner runner)
     {
-        Stream<Map<PlanNodeId, PlanNodeStats>> stagesStatsStream =
-                getAllStages(Optional.of(outputStageInfo)).stream()
-                        .map(PlanNodeStatsSummarizer::aggregatePlanNodeStats);
-
-        Map<PlanNodeId, PlanNodeStats> mergedStats = mergeStats(stagesStatsStream);
-        return transformValues(mergedStats, MetricComparator::toPlanNodeStats);
+        return transaction(runner.getTransactionManager(), runner.getAccessControl())
+                .singleStatement()
+                .execute(runner.getDefaultSession(), session -> {
+                    String queryId = runner.executeWithQueryId(session, query).getQueryId();
```

Comment:
See `Add createPlan() method to QueryRunner interface` & `Support LocalQueryRunner in StatisticsAssertion` (they appear on GH out of order, sorry!) -- In the case of `LocalQueryRunner`, we just plan the query. In the case of `DistributedQR`, we don't have such a REST API, this is why we execute query (and we cancel it immediately, since `Add createPlan() method to QueryRunner interface` commit). Tests using `LocalQueryRunner` are indeed faster, but I don't think it's a big deal. Also, i didn't profile to know what cases biggest overhead.

Reply:
... Now I think the biggest overhead is aggregation queries computing actual values to compare stat estimations with. I.e. not something that can be fixed.

Reply:
ok. If it's not causing problems, no need to worry about it.